### PR TITLE
CSSOM: Whole ton of things

### PR DESCRIPTION
**Pull request by Manishearth**
CSSOM is now starting to be useful!

Based on #14190. Only the <s>last commit</s> last two commits need review.

cc @xidorn . This doesn't change the style API, but adds useful methods.

part of #11420

This adds:
 - `insertRule()` and `deleteRule()` on `CSSStyleSheet`, `CSSGroupingRule`
 - `.style` getters on link and style elements
 - Keyframes-backed `CSSRules` and `CSSKeyframesRule.cssRules`
 - `CSSGroupingRule.cssRules`
 - `prefix` and `namespaceURI` attributes of `CSSNamespaceRule`
 - Fixups regarding parent stylesheets

r? @SimonSapin

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14241)

### CODE REVIEW

File: components/script/dom/cssrulelist.rs [hunk 1]
```diff
@@ -40,12 +45,83 @@ impl CSSRuleList {
...
+            }
+        }
+        let global = self.global();
+        let window = global.as_window();
+        let doc = window.Document();
+        let index = idx as usize;
+
+        // Step 1, 2
+        // XXXManishearth get url from correct location
+        // XXXManishearth should we also store the namespace map?
```
Comment (Manishearth) [inline components/script/dom/cssrulelist.rs]: I'm considering just filing issues for these

**Comment by Manishearth** [comment]
Also http://build.servo.org/builders/linux-rel-wpt/builds/862

**Comment by SimonSapin** [comment]
Looks good modulo inline comments. (Opening this before #14190 is merged makes review slightly harder…)

---

Review status: 0 of 75 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/script/dom/cssrulelist.rs, line 79 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWeY2gyaTzL9GRb0UA--r6-79:-KWi_Nczp3pByd1mf5n5:b-f1l8qa) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/script/dom/cssrulelist.rs#L79)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> spec is very vague here. will have to check against gecko (and file an issue)
</details>
The rules are:

* `@charset`, if present, must be first (but it’s not present in the OM so don’t worry about it)
* `@import` must be before everything other than `@charset` https://drafts.csswg.org/css-cascade/#at-import
* `@namespace` must be before everything other than `@charset` or `@import` https://drafts.csswg.org/css-namespaces/#syntax
* None of the above can appear inside a conditional rule (`@media` or `@supports`) https://drafts.csswg.org/css-conditional/#contents-of

Tracking this is what the split between `TopLevelRuleParser` and `NestedRuleParser` as well as the `TopLevelRuleParser.state` field are for. As mentioned in another comment, some things probably need to be moved around for this.

---

*[components/script/dom/cssrulelist.rs, line 82 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWibcLIT4ms9yC0emDv:-KWibcLIT4ms9yC0emDw:b-anq4pb) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/script/dom/cssrulelist.rs#L82)):*
> ```Rust
> 
>             // Step 6
>             if let StyleCssRule::Namespace(..) = new_rule {
> ```

Something implicit in CSSOM is that namespace prefixes in selectors in a stylesheet are based on all `@namespace` rules in the stylesheet, even inserted by CSSOM. So this code will need to update the `NamespaceMap` that is currently only maintained during stylesheet parsing but will need to be kept in the `Stylesheet` struct.

(This is why this `InvalidStateError` exception exists: so that selectors don’t need to be re-parsed.)

---

*[components/script/dom/cssrulelist.rs, line 115 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWiYLH9w3noPgTZnCvE:-KWiYLHApS-r4C8l9bWP:b5ehsde) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/script/dom/cssrulelist.rs#L115)):*
> ```Rust
>         let mut dom_rules = self.dom_rules.borrow_mut();
>         self.rules.0.write().remove(index);
>         dom_rules[index].get().map(|r| r.disown());
> ```

Nit: `Vec::remove` returns the removed element, `disown` can be called on that.

---

*[components/style/stylesheets.rs, line 136 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWiQ4tnhNkmkorP1KY7:-KWiQ4tocnLFyrerpF6K:b-wy09fo) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/style/stylesheets.rs#L136)):*
> ```Rust
>     }
> 
>     pub fn from_str(css: &str, origin: Origin,
> ```

Nit: I’d prefer to call this `parse`.

---

*[components/style/stylesheets.rs, line 139 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWiPzJe8GH1Oge810ud:-KWiPzJe8GH1Oge810ue:b-deewj0) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/style/stylesheets.rs#L139)):*
> ```Rust
>                     base_url: Url, extra_data: ParserContextExtraData) -> Result<Self, ()> {
>         let error_reporter = Box::new(MemoryHoleReporter);
>         let rule_parser = TopLevelRuleParser {
> ```

`TopLevelRuleParser` is what you want for `CSSStyleSheet.insertRule`, but for `CSSGroupingRule.insertRule` it’s `NestedRuleParser`. Probably these two Rust types should be merged, with a boolean field or something to tell both cases apart. This boolean would be an additional parameter to this method.

If you’d rather not do this now I can do it in a follow-up PR.

---

*[components/style/stylesheets.rs, line 145 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWiNzRExP2D2KOU8bPh:-KWiNzRExP2D2KOU8bPi:b-5d7yso) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/style/stylesheets.rs#L145)):*
> ```Rust
>         };
>         let mut input = Parser::new(css);
>         let mut iter = RuleListParser::new_for_stylesheet(&mut input, rule_parser);
> ```

This should use `cssparser::parse_one_rule` instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14241#-:-KWiQjmg5I4gysRBgTQS:b-fxzxme)*

**Comment by Manishearth** [comment]
Review status: 0 of 75 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/script/dom/cssrulelist.rs, line 115 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWiYLH9w3noPgTZnCvE:-KWj2jZ8BA7El2nrP-7R:b-iwgewm) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/script/dom/cssrulelist.rs#L115)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary>
> Nit: `Vec::remove` returns the removed element, `disown` can be called on that.
</details>
I'm wary of GC happening within that step.

---

*[components/style/stylesheets.rs, line 139 at r6](https://reviewable.io:443/reviews/servo/servo/14241#-KWiPzJe8GH1Oge810ud:-KWj5LF5oXDVpo-4qTW-:bqpw53p) ([raw file](https://github.com/servo/servo/blob/3293fd0a4cbcc7a0753adc8daa56174695fb85d1/components/style/stylesheets.rs#L139)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary>
> `TopLevelRuleParser` is what you want for `CSSStyleSheet.insertRule`, but for `CSSGroupingRule.insertRule` it’s `NestedRuleParser`. Probably these two Rust types should be merged, with a boolean field or something to tell both cases apart. This boolean would be an additional parameter to this method.
> 
> If you’d rather not do this now I can do it in a follow-up PR.
</details>
I didn't merge them, I just changed CssRule::parse to accept a nested parameter

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14241)*

**Comment by Manishearth** [comment]
Made it handle state (and other review fixups)

**Comment by Manishearth** [comment]
Added support for keyframe-backed `CSSRuleList`s, and `CSSKeyframeRule`. No insertion/deletion from `CSSKeyframesRules` yet (but it's easy to implement).

**Comment by Manishearth** [comment]
I'm going to wait for @canaltinova's ToCSS work (at #14238) to land before running this through try again since having working serialization may make more tests pass.

**Comment by SimonSapin** [comment]
Reviewed 4 of 5 files at r6, 12 of 12 files at r7, 5 of 5 files at r8, 15 of 15 files at r9, 3 of 3 files at r10, 8 of 8 files at r11, 11 of 11 files at r12, 4 of 4 files at r13, 4 of 4 files at r14, 7 of 7 files at r15, 3 of 3 files at r16, 1 of 1 files at r17, 3 of 3 files at r18.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/cssstylesheet.rs, line 61 at r7](https://reviewable.io:443/reviews/servo/servo/14241#-KXBM_Ze41Mg7oujfKav:-KXBM_ZfDAObiGUU_t8Z:buj252n) ([raw file](https://github.com/servo/servo/blob/d9db66ecab4387660a0585d9b1d6e936e6d2efa1/components/script/dom/cssstylesheet.rs#L61)):*
> ```Rust
>     // https://drafts.csswg.org/cssom/#dom-cssstylesheet-insertrule
>     fn InsertRule(&self, rule: DOMString, index: u32) -> Fallible<u32> {
>         // XXXManishearth check origin clean flag
> ```

File an issue for this?

---

*[components/style/stylesheets.rs, line 68 at r10](https://reviewable.io:443/reviews/servo/servo/14241#-KXBP7Lld5wJB8aQy6AK:-KXBP7Lld5wJB8aQy6AL:b-d6rnfe) ([raw file](https://github.com/servo/servo/blob/60bb89030f29eaaebb8efdf8f6291051b6c9cce2/components/style/stylesheets.rs#L68)):*
> ```Rust
>     pub fn state_at_index(rules: &[CssRule], at: usize) -> State {
>         let mut state = State::Start;
>         for rule in rules.iter().take(at) {
> ```

Does this need to be a loop at all? Everything done in previous iterators is overwritten in the last iterator. (Same for `state_at_index_rev` below.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14241)*

**Comment by Manishearth** [comment]
Addressed

**Comment by SimonSapin** [comment]
@bors-servo r+

---

Reviewed 2 of 2 files at r19.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14241#-:-KXGt4pzYeDC_vtMAxyJ:bkg9nno)*

**Comment by SimonSapin** [comment]
```rust
error[E0063]: missing fields `namespaces`, `stylesheet_origin` in initializer of `stylesheets::TopLevelRuleParser<'_>`
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/stylesheets.rs:196:31
    |
196 |         let mut rule_parser = TopLevelRuleParser {
    |                               ^^^^^^^^^^^^^^^^^^ missing `namespaces`, `stylesheet_origin`
```

Ah, some of your new code needs to change because I’ve since added fields in another PR. Sorry!

**Comment by upsuper** [comment]
You didn't successfully cc'ed me...

The algorithm of `insertRule` and `deleteRule` really should be implemented inside style component so that they can be reused for stylo...

**Comment by Manishearth** [comment]
I tried to make it as reusable as possible. Because of the parallel dom tree the algorithm involves DOM code as well and is hard to separate. The parsing/insertion validity algorithms are in stylesheets.rs though.

**Pull request merged**